### PR TITLE
update scene

### LINKED ISSUE #4: Scene
Methods: create/update/destroy:

- **mesh** (vertices - float3 + indices - uint32): (SoA, AoS)

> v0, v1, v2 …. [v120 - v160] ..  - vertex buffer (vertex attribs: float4 pos, float2 uv, float3 normal)

> 0, 1, 2.. [305 - 700]  - index buffer

- **material** (float4 - color)
- **instance** (mesh id + transform + material)
- **light** (float3-4)

### PR DESCRIPTION
**Pull request by DoktaPola**

### CODE REVIEW

**Comment by DoktaPola** [comment]
1. Removed functions 'add' & 'remove'. 
2. Made 3 func-s : removeMesh/Instance/Material to delete one of this objects.
3. Set up clang-format

**Comment by DoktaPola** [comment]
1) added several doctests (and comment some which need to be completed);
2) added "void updateInstanceTransform(uint32_t instId, glm::mat4 newTransform)" method;
3) added  "void beginFrame(), void endFrame()" methods;
4) changed "remove instance/mesh/material" methods as well as " create instance/mesh/material";

File: src/scene/scene.cpp [hunk 1]
```diff
@@ -16,27 +32,122 @@ uint32_t Scene::createMesh(const std::vector<Vertex>& vb, const std::vector<uint
     }
     // copy vertices
     mVertices.insert(mVertices.end(), vb.begin(), vb.end());
+    return meshId;
+}
 
-    // add mesh to storage
-    mMeshes.push_back(mesh);
+void MeshInstance::init_rotateBy(const float& degrees)
```
Comment (ikryukov) [inline src/scene/scene.cpp]: remove

File: src/scene/scene.cpp [hunk 2]
```diff
@@ -16,27 +32,122 @@ uint32_t Scene::createMesh(const std::vector<Vertex>& vb, const std::vector<uint
...
+{
+    this->rotateBy(degrees);
+}
 
-    const uint32_t meshId = mMeshes.size() - 1;
-    return meshId;
+glm::mat4 MeshInstance::getTransformMatrix() const
```
Comment (ikryukov) [inline src/scene/scene.cpp]: float4x4& <- ref

File: src/scene/camera.cpp [hunk 3]
```diff
@@ -2,159 +2,172 @@
...
+    {
+        matrices.view = transM * rotM;
+    }
+    updated = true;
 }
 
 bool Camera::moving()
 {
-	return keys.left || keys.right || keys.up || keys.down;
+    return keys.left || keys.right || keys.up || keys.down;
```
Comment (ikryukov) [inline src/scene/camera.cpp]: tabs? use clang-format

**Review by ikryukov — CHANGES_REQUESTED** [review]

File: include/scene/scene.h [hunk 4]
```diff
@@ -1,32 +1,71 @@
...
-    glm::vec2 uv;
-};
 
 struct Mesh
 {
     uint32_t mIndex; // Index of 1st index in index buffer
     uint32_t mCount; // amount of indices in mesh
 };
 
+struct MeshInstance
```
Comment (ikryukov) [inline include/scene/scene.h]: remove: duplicate of Instance

File: src/scene/camera.cpp [hunk 5]
```diff
@@ -1,160 +1,173 @@
...
+    {
+        // todo: move code from example base class for look-at
+    }
+
+    if (retVal)
+    {
+        updateViewMatrix();
+    }
+    return retVal;
+}
```
Comment (ikryukov) [inline src/scene/camera.cpp]: new line

File: src/scene/camera.cpp [hunk 6]
```diff
@@ -1,160 +1,173 @@
...
-	if (retVal) {
-		updateViewMatrix();
-	}
-	return retVal;
-}
+#include "camera.h"
+
+void Camera::updateViewMatrix()
+{
+    glm::mat4 rotM = glm::mat4(1.0f);
```
Comment (ikryukov) [inline src/scene/camera.cpp]: mat3 -> float4x4

File: tests/scene/tests_scene.cpp [hunk 7]
```diff
@@ -1,9 +1,129 @@
...
+    CHECK(scene.mInstances.size() == 1);
+}
+
+TEST_CASE("test checkMaterial")
+{
+    nevk::Scene scene;
+    uint32_t matId = scene.createMaterial(glm::float4(1.0));
+    CHECK(matId == 0);
+    CHECK(scene.mMaterials.size() == 1);
+}
```
Comment (ikryukov) [inline tests/scene/tests_scene.cpp]: new line

File: .gitignore [hunk 8]
```diff
@@ -2,4 +2,4 @@ build/
 *.exe
 .vs
 .idea/s
-.DS_Store
+.DS_Store
```
Comment (ikryukov) [inline .gitignore:5]: new line

**Review by ikryukov — CHANGES_REQUESTED** [review]

File: src/scene/scene.cpp [hunk 9]
```diff
@@ -1,42 +1,131 @@
...
+    const uint32_t ibOffset = mIndices.size();  // adjust indices for global index buffer
+    for (int i = 0; i < ib.size(); ++i)
+    {
+        mIndices.push_back(ibOffset + ib[i]);
+    }
+    mVertices.insert(mVertices.end(), vb.begin(), vb.end());  // copy vertices
+    return meshId;
+}
+
+glm::mat4 Scene::createMeshTransform()
```
Comment (ikryukov) [inline src/scene/scene.cpp]: этот метод не пригодится

File: tests/scene/tests_scene.cpp [hunk 10]
```diff
@@ -1,9 +1,129 @@
...
-    CHECK(e != nullptr);
-}
+#include <scene/scene.h>
+
+#include <doctest.h>
+
+
+TEST_CASE("test sceneCreation")
+{
+    auto* scene = new nevk::Scene();
```
Comment (ikryukov) [inline tests/scene/tests_scene.cpp]: тест не проверяет логику

File: include/scene/scene.h [hunk 11]
```diff
@@ -1,78 +1,129 @@
...
+    {
+        glm::float3 pos;
+        glm::float3 normal;
+        glm::float2 uv; // hold the texture coordinate
+        glm::float3 color;
+    };
+
+    bool fr_mod{};
+
+    std::set<uint32_t> mDirtyInstances;
```
Comment (ikryukov) [inline include/scene/scene.h]: эти все поля в private

**Review by ikryukov — CHANGES_REQUESTED** [review]

File: include/scene/scene.h [hunk 12]
```diff
@@ -33,9 +30,18 @@ struct Instance
...
+    std::stack<uint32_t> mDelMesh;
+    std::stack<uint32_t> mDelMaterial;
+
+    bool FrMod{};
+
+    std::set<uint32_t> mDirtyInstances;
 
 public:
-    struct Vertex
+      struct Vertex
```
Comment (ikryukov) [inline include/scene/scene.h:44]: прокрался пробел

**Review by ikryukov — CHANGES_REQUESTED** [review]

**Pull request merged**